### ISSUE TITLE
Exposed ABI depends on C++ standard used, while it could be avoided

### ISSUE THREAD

**Issue by 0xd34df00d** [arvidn/libtorrent#3602]
tl;dr: libtorrent built with C++11 is not compatible with revdeps built with C++≥14 and vice versa

For now I've identified only one culprit: https://github.com/arvidn/libtorrent/blob/30b8a06bd7999f3940633acac28f51cc2a0471a2/include/libtorrent/entry.hpp#L90

If built with `-std=c++11`, libtorrent exposes `libtorrent::entry::dict() const` symbol (among the others). If built with `-std=c++14` or `-std=c++17`, it exposes `libtorrent::entry::dict[abi:cxx11]() const` instead. And, analogously, if some client code dependent on libtorrent is built with C++11, then it'll expect the former version, and the latter in C++14/17 mode.

I see two ways to solve this off the top of my head:
1. Document this behaviour and make sure that all the revdeps are built with consistent set of flags (which is not always easily possible for linux distros).
2. Expose a configure option that'll force falling back to the C++11 behaviour, and expose it in `.pc` so that client code could be aware.

@arvidn what are your thoughts on this?

**Comment by arvidn** [comment]
I don’t think there’s a simple solution. I think the best option is just to make sure there’s a link error by encoding the version in an inline namespace.

**Comment by 0xd34df00d** [comment]
That's a nice workaround! Unfortunately, it will break if libtorrent itself is used from a .so — in this case the linker will complain at runtime (and, quite possibly, at `dict` usage time).

BTW there's a third option — just sticking with C++11-compatible implementation until libtorrent drops support for C++11. That's perhaps a bit ugly, but solves all of the problems. What do you think?

**Comment by arvidn** [comment]
It is a pessimisation, as it requires constructing a string for every lookup.
